### PR TITLE
Add clipping planes to `VolumeVisual`

### PR DESCRIPTION
**Pull request by brisvag**
After trying to implement this [in napari](https://github.com/napari/napari/pull/2934) and getting some very useful input in #2113, here's a pure vispy implementation of clipping planes for the Volume visual.

I added an example (`examples/basics/scene/volume_clipping.py`) with some simple keybindings to showcase the code (thanks @alisterburt for your code at #2074 :)).

https://user-images.githubusercontent.com/23482191/123859348-d1646600-d924-11eb-8c12-d28f9fe087a9.mp4

Planes are defined as two vec3 (position and normal vector), and are used to clip the volume (anything on the negative side of the normal vector is not shown). The shape of `volume.clipping_planes` should therefore be `(n_planes, 2, 3)`.

Currently, the vector coordinates are normalized in (0, 1), but they should probably be in world coordinates; I'm not sure how to handle that...

### CODE REVIEW

**Comment by djhoese** [comment]
Cool! That was quick. If it is ok with you and @alisterburt I'd like to try to get this into VisPy 0.8 along with #2074. VisPy 0.7 should be this week if we can come to a consensus on some other PRs and then we'll work on 0.8 for a couple weeks to a month after that if these PRs get finished.

Currently it looks like there are some linting complaints about the code. Otherwise, I'll try to review this when I can find the time (probably after 0.7 is released).

**Comment by brisvag** [comment]
That's fine by me! If we're ever in a rush, we can always vendor this on the napari side.

**Review by alisterburt — CHANGES_REQUESTED** [review]
I've left a PR at brisvag/vispy#1 which fixes a few little problems with this PR, namely
- the clipping planes should be defined in data coordinates rather than texture coordinates
- the axis ordering for clipping planes should be consistent with the input data volume

There are a few more tiny issues I think should be addressed before merge
- clearer docstring for definition of clipping planes, specifying the coordinate system in which the planes are defined and providing an example
- there should be some validation of the input data for the clipping planes attribute. If the data doesn't have the right shape currently this will fall over in all sorts of horrible ways

I think we should write some documentation about how this works too so it's a bit easier to get our heads around if we or others need to maintain it in the future
- at minimum a comment in the shader next to where your function is injected
- ideally something for the vispy docs which explains this aspect of volume rendering

Otherwise this is just incredible!! Awesome job and congrats on such a cool feature - I'm really excited to see what we can do with this 😍

**Comment by djhoese** [comment]
Good ideas for documentation. I think I need to update the website again and define some areas for documentation like this. Like a place specifically for how Volume rendering is performed that could include screenshots. Same idea could be used for Markers and Image and some of the others. It might be a little too ugly to shove all of this information into the docstrings and require a user to look at the API docs.

**Comment by alisterburt** [comment]
@djhoese happy to help with a documentation push! @almarklein's [blog post](https://almarklein.org/volume_rendering.html) was instrumental when @kevinyamauchi and I were trying to get our heads around the volume rendering shader - we could probably reuse a bunch of what he's done there if we ask nicely

**Comment by brisvag** [comment]
> I've left a PR at [brisvag#1](https://github.com/brisvag/vispy/pull/1) which fixes a few little problems with this PR, namely

Sweet, that's the stuff I was not sure how to deal with! I'll check it out.

>     * there should be some validation of the input data for the clipping planes attribute. If the data doesn't have the right shape currently this will fall over in all sorts of horrible ways

This was a deliberate choice: I noticed that basically nothing is validated at this level except for the data, so I assumed that it porbably happens at some point at higher levels. If that's not the case, then other stuff could benefit from some validation as well, otherwise failures show up in glsl gibberish form rather than beautiful python introspection :P

> I think we should write some documentation about how this works too so it's a bit easier to get our heads around if we or others need to maintain it in the future

Yeah, good idea!

File: vispy/visuals/volume.py
Comment (almarklein) [inline vispy/visuals/volume.py:252]: I wonder if we can avoid having this code inside the loop, and instead modify the `start_loc` and `nsteps` of the ray.

Comment (almarklein) [reply vispy/visuals/volume.py:252]: Mmm, I guess that could add a lot of complexity too. Alternatively, could we (or do we?) avoid this check if there are no clipping planes?

**Comment by almarklein** [comment]
> we could probably reuse a bunch of what he's done there if we ask nicely

Of course!

File: vispy/visuals/volume.py
Comment (brisvag) [reply vispy/visuals/volume.py:252]: That wasmy first thought as well, but then I thought that if the clipped volume is convex, it wouldn't work - which could happen with my original idea of having the ability to do both intersections and unions of clipped volumes :P. However, that's more complicated than I thought (you'd want to set up an expression to determine the order of unions and intersections...), so maybe moving this to the vertex shader is the way to go?

The check is currently not skipped if there are no clipping planes; the function simply returns 1.0.

Comment (almarklein) [reply vispy/visuals/volume.py:252]: > if the clipped volume is concave, it wouldn't work 

Ah, I see. Interesting idea to do logic ops with the clipped volume!

> so maybe moving this to the vertex shader is the way to go?

You mean the logic for determining whether a loc is visible or not, right? I'm not in deep enough to see what makes it complicated. Would it make sense to do the complicated stuff in Python and generate the code for `clip_by_planes` accordingly?

Comment (brisvag) [reply vispy/visuals/volume.py:252]: Yes, I mean that. I'm not sure either, I'm very new to glsl and shaders... I don't quite understand what the vertex shader does in the context of volumes, or how I would go from the `start_loc` to a `clipped_start_loc`.
We're limited on how much we can generate from python, because everything has to happen before compile time, so I'm not sure how much that can help.

**Comment by brisvag** [comment]
I've made some updates to fix a few issues and bring it up to date with what changed in #2117 and such.

Functionality is the same, but a bit more solid (passing `None` now works without issues). I added some caching so that the code is computed only once for each amount of planes.

File: vispy/visuals/volume.py
Comment (djhoese) [inline vispy/visuals/volume.py:711]: There should probably be some upper limit on this.

File: vispy/visuals/volume.py
Comment (djhoese) [inline vispy/visuals/volume.py:715]:
```suggestion
        """Build the code snippet used to clip the volume based on self.clipping_planes."""
```

File: vispy/visuals/volume.py
Comment (djhoese) [inline vispy/visuals/volume.py:743]:
```suggestion
        """Get the set of planes used to clip the volume.
        
        Each plane is defined by a position and a normal vector (magnitude is irrelevant). Shape: (n_planes, 2, 3)
        """
```

**Review by djhoese — CHANGES_REQUESTED** [review]
Small docstring changes.

Also it looks like the tests are failing because Pillow is failing to install from the wheel (missing zlib). I'm not sure why that is only just starting to fail as Pillow hasn't had a release in a month.

Also, could you put your example in `examples/scene/` (you'll need to `main` if the directory doesn't exist for you).

File: vispy/visuals/volume.py
Comment (brisvag) [reply vispy/visuals/volume.py:711]: What would you consider reasonable? Keep in mind that this caches a new result depending on the *number* of clipping planes, not what the clipping planes look like. I don't expect anyone to reasonably go above 10 clipping planes, so we can cache about 10?

File: examples/basics/scene/volume_clipping.py
Comment (djhoese) [inline examples/basics/scene/volume_clipping.py:8]:
```suggestion
# -*- coding: utf-8 -*-
# vispy: gallery 5
# -----------------------------------------------------------------------------
# Copyright (c) Vispy Development Team. All Rights Reserved.
# Distributed under the (new) BSD License. See LICENSE.txt for more info.
# -----------------------------------------------------------------------------
"""
Example of volume rendering with clipping planes
================================================

Controls:
```

File: vispy/visuals/volume.py
Comment (djhoese) [reply vispy/visuals/volume.py:711]: I guess 10 works for me. I could see less, but I am not the targeted user for this so I'm not sure. Just wanted to avoid a "memory leak" by leaving this unbounded. 10 works.

**Comment by brisvag** [comment]
Updated what requested. The pillow issue is strange indeed...

**Comment by djhoese** [comment]
Ah the wheel building is failing on Python 3.10 which must have just been added to cibuildwheel.

@almarklein @kmuehlbauer Any idea how you want to work around this? We could not build for 3.10 and then regret it later if/when we forget. Or we could merge knowing that these wheel builds will fail. Or we can wait until Pillow releases a wheel for Python 3.10. Or we could install the necessary dependency libraries so Pillow can build from source.

**Comment by almarklein** [comment]
> Any idea how you want to work around this? 

Can we release the 3.10 wheel but not test that particular one (for now)?

**Comment by djhoese** [comment]
We could try this: https://cibuildwheel.readthedocs.io/en/stable/options/#test-skip

**Comment by brisvag** [comment]
Now that 2169 is merged, all tests pass. Can this be merged?

**Comment by djhoese** [comment]
Yes. It is on my TODO list today to review all the PRs that should be ready for merging. Doing a few other things first.

**Comment by alisterburt** [comment]
thanks @djhoese - really appreciate you taking the time! 

I have quite a bit more free time than usual this week so don't hesitate to ask for things you think are important for this and the plane rendering PR, I'm happy to sink time in to get these right

File: examples/scene/volume_clipping.py
Comment (djhoese) [inline examples/scene/volume_clipping.py:8]:
```suggestion
Volume rendering with clipping planes
```

File: examples/scene/volume_clipping.py
Comment (djhoese) [inline examples/scene/volume_clipping.py:9]:
```suggestion
Volume rendering with clipping planes
=====================================
```

**Review by djhoese — APPROVED** [review]
This makes enough sense to me...I think. It also seems to involve less changes that the rendering planes so maybe any conflicts will be easier to resolve if we merge this first.

**Comment by brisvag** [comment]
Sweet! Thanks for reviewing @djhoese, looking forward to more contributions soon :)

**Comment by alisterburt** [comment]
Thanks @kmuehlbauer ! Feels great to  help close such a long-standing issue 🙂

**Pull request merged**
